runtime/runtime/src/congestion_control.rs

```diff
@@ -461,9 +521,24 @@ impl DelayedReceiptQueueWrapper {
     }
 }
 
-pub(crate) fn receipt_size(receipt: &Receipt) -> Result<usize, IntegerOverflowError> {
-    // `borsh::object_length` may only fail when the total size overflows u32
-    borsh::object_length(&receipt).map_err(|_| IntegerOverflowError)
+/// Get the receipt size from the receipt that was retrieved from the state.
+/// If it is a [Receipt], the size will be computed.
+/// If it s the [StateStoredReceipt], the size will be read from the metadata.
+pub(crate) fn receipt_size(
+    receipt: &ReceiptOrStateStoredReceipt,
+) -> Result<u64, IntegerOverflowError> {
+    match receipt {
+        ReceiptOrStateStoredReceipt::Receipt(receipt) => compute_receipt_size(receipt),
+        ReceiptOrStateStoredReceipt::StateStoredReceipt(receipt) => Ok(receipt.metadata().size),
+    }
+}
+
+/// Calculate the size of a receipt before it is pushed into a state queue or
+/// buffer. Please note that this method should only be used when storing
+/// receipts into state. It should not be used for retrieving receipts from the state.
+pub(crate) fn compute_receipt_size(receipt: &Receipt) -> Result<u64, IntegerOverflowError> {
```

Comment:
Splitting `receipt_size` into `compute_receipt_size` and `receipt_size` is very nice. Makes it much easier to make sure that computing happens only in places where it should :+1: